File: instrumentation/kamon-annotation-api/src/main/java/kamon/annotation/api/Gauge.java
/*
 * =========================================================================================
 * Copyright © 2013-2015 the kamon project <http://kamon.io/>
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 * =========================================================================================
 */

package kamon.annotation.api;

import java.lang.annotation.*;

/**
 * A marker annotation to define a method as a Gauge.
 * <p>
 * <p>
 * Given a method like this:
 * <pre><code>
 *     {@literal @}Gauge(name = "coolName", tags="${'my-cool-tag':'my-cool-operationName'}")
 *     public (Long|Double|Float|Integer) coolName() {
 *         return 100L;
 *     }
 * </code></pre>
 * <p>
 * <p>
 * A {@link kamon.metric.Gauge Gauge} for the defining method with the name {@code coolName} will be created and each time the
 * {@code #coolName(String)} method is invoked, the gauge will be set the returned value.
 */
@Documented
@Target(ElementType.METHOD)
@Retention(RetentionPolicy.RUNTIME)
public @interface Gauge {

    /**
     * @return The gauge's name.
     * <p>
     * Also, the Metric name can be resolved with an EL expression that evaluates to a String:
     * <p>
     * <pre>
     * {@code
     *  class Gauged  {
     *        private long id;
     *
     *        public long getId() { return id; }
     *
     *        {@literal @}Gauge (name = "${'counterID:' += this.id}")
     *        void gaugedMethod() {} // create a gauges with name => gaugeID:[id]
     *    }
     * }
     * </pre>
     */
    String name() default "";

    /**
     * Tags are a way of adding dimensions to metrics,
     * these are constructed using EL syntax e.g. "${'algorithm':'1','env':'production'}"
     *
     * @return the tags associated to the counter
     */
    String tags() default "";
}

File: core/kamon-core/src/main/scala/kamon/metric/Gauge.scala
/*
 * Copyright 2013-2021 The Kamon Project <https://kamon.io>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kamon.metric

import java.util.concurrent.atomic.AtomicLong
import java.util.function.LongUnaryOperator

import kamon.metric.Metric.{BaseMetric, BaseMetricAutoUpdate}
import kamon.tag.TagSet
import java.lang.Double.{longBitsToDouble, doubleToLongBits}

/**
  * Instrument that tracks the latest observed value of a given measure.
  */
trait Gauge extends Instrument[Gauge, Metric.Settings.ForValueInstrument] {

  /**
   * Increments the current value by one.
   */
  def increment(): Gauge

  /**
   * Increments the current value the provided number of times.
   */
  def increment(times: Double): Gauge

  /**
   * Decrements the current value by one.
   */
  def decrement(): Gauge

  /**
   * Decrements the current value the provided number of times.
   */
  def decrement(times: Double): Gauge

  /**
    * Sets the current value of the gauge to the provided value.
    */
  def update(value: Double): Gauge

}

object Gauge {

  /**
    * Gauge implementation backed by a volatile variable.
    */
  class Volatile(val metric: BaseMetric[Gauge, Metric.Settings.ForValueInstrument, Double], val tags: TagSet)
      extends Gauge
      with Instrument.Snapshotting[Double]
      with BaseMetricAutoUpdate[Gauge, Metric.Settings.ForValueInstrument, Double] {

    // https://stackoverflow.com/questions/5505460/java-is-there-no-atomicfloat-or-atomicdouble
    private val _currentValue = new AtomicLong(0)

    override def increment(): Gauge = increment(1)

    override def increment(times: Double): Gauge = {
      _currentValue.updateAndGet(new LongUnaryOperator {
        override def applyAsLong(v: Long): Long = {
          val newValue: Double = longBitsToDouble(v) + times
          if (newValue < 0) v else doubleToLongBits(newValue)
        }
      })
      this
    }

    override def decrement(): Gauge = decrement(1)

    override def decrement(times: Double): Gauge = {
      _currentValue.updateAndGet(new LongUnaryOperator {
        override def applyAsLong(v: Long): Long = {
          val newValue: Double = longBitsToDouble(v) - times
          if (newValue < 0) v else doubleToLongBits(newValue)
        }
      })
      this
    }

    override def update(newValue: Double): Gauge = {
      if (newValue >= 0d)
        _currentValue.set(doubleToLongBits(newValue))

      this
    }

    override def snapshot(resetState: Boolean): Double =
      longBitsToDouble(_currentValue.get())

    override def baseMetric: BaseMetric[Gauge, Metric.Settings.ForValueInstrument, Double] =
      metric
  }
}
